### DISCUSSION TITLE
Chat (Chain) Logging

### DISCUSSION THREAD

**Discussion by matthias** [FlowiseAI/Flowise#492] [Ideas]
The idea is to have an UI in Flowise to review (external) chats.
As a minimum: Input / Output - but the more context we can provide - the better

What's already implemented on some chains is to use verbose logging - to the console.

I digged into this - and langchain provides a mechansim to hook into the chain(flow):
https://js.langchain.com/docs/production/callbacks/create-handlers

The idea would be to write a custom CallbackHandler and add it to the LLM call (in the Flowise Chain** components)

This should extract useful info and write it to the DB - so we can show it in an UI (list of past chats). For now I haven't thought too much about privacy - but maybe something like an option / parameter on the external chat api could switch this on / off.

It's for sure useful to learn from the inputs / generated outputs - on the various chains.

Just to let you know we are actively working on this - and if you have any plans / ideas in that direction I appreciate the dialogue.

**Comment by marius-smg** [comment]
This would be super helpfull. I'm developing Bots for my customers and they want an analysis of the usage data and atm i don't have an idea how to implement this myself. what would be great, is if the data would be separated by chatflow instance. This way you know what bot is used most, what questions are asked to what bot and so on.

**Comment by HenryHengZJ** [comment]
we already have custom chain handler - https://github.com/FlowiseAI/Flowise/blob/main/packages/components/src/utils.ts#L207

just needed to find a way to present the info to the UI

**Reply by matthias** [reply]
Should we reuse / extend this one. As far as I understand the CustomChainHandler serves a different purpose. And is only used *if* isStreaming. So I guess we should create a new Handler just for Logging.

**Reply by HenryHengZJ** [reply]
created another handler for logger- [PR](https://github.com/FlowiseAI/Flowise/pull/518)
